### ISSUE TITLE
Case sensitivity with cookie paths

### ISSUE THREAD

**Issue by MH61Aus** [DuendeSoftware/Support#1202]
**Which version of Duende IdentityServer are you using?**

7

**Which version of .NET are you using?**

8

**Describe the bug**

Question - Is there a recommended practice when it comes to case sensitivity and cookies? 

For example, with an instance deployed to Domain/Identity, the /Identity path is case sensitive in the cookies.

We ran into this with our setup of 2 web clients and a console client. The Identity Server uses asp.net Identity integration, so it has a profile page, etc. In our pipelines, we set up our instances with domain/Identity plug these same values into our appsettings.
We noticed with the console client that when running client.GetDiscoveryDocumentAsync(identityServerUri); it failed because of /Identity in the URI, so we added a ToLower when reading the config.

I was working on webclient1 at the same time, so I figured I'd also add a ToLower when reading the config, which probably wasn't necessary since setting the Authority of OIDC doesn't seem to care about the case.

But, Webclient 1 also has a link to the account management page on the identity server. we implemented it in "this will open your profile in a new tab" style. We noticed that this was fine on local, but on our other environments, the user was asked for their credentials again. 

Webclient 2 never did the ToLower on the authority. So if you navigate form webclient 1 to webclient 2, on any environment other than localhost you need to enter credentials again.

so basically SSO isn't working.

Looking at the cookies we can see that its all to do with case mismatch on the path in the cookie. here's a screenshot where i've created a more extreme/obvious example with the casing in the path.
![image](https://github.com/DuendeSoftware/Support/assets/34496310/5c27f687-52a5-4963-bcc2-8378256e5cc5)

So, we should be able to get around this by ensuring that any time we generate any uri for identity (i.e. setting the authority or linking to account management page) we do a toLower first. But I guess I'm trying to figure out if this is all by design? I'm surprised I'm not seeing more instances of people running into this behaviour. I'm surprised i'm not seeing this in some sort of "common deployment issues" or SSO doc.

**Comment by AndersAbel** [comment]
Both the cookie path and the OpenID Connect issuer are case sensitive strings. This is specified in the cookie and OIDC standards.

I would recommend that you one casing throughout all of your configuration in both the IdentityServer host and client applications, wouldn't that solve the issue? Or am I missing something here?

We do not generally recommend to host the IdentityServer beneath a path. The security model for cookie isolation works better if IdentityServer is hosted on a separate host. If you want/need to host IdentityServer beneath a path I would recommend to set up the web server to be case sensitive in what it accepts. If IdentityServer answers to different casing of the path, it might issue tokens with different casing to the issuer field which would cause validation problems.

I would also like to point out that the `idsrv.session` cookie is not the main session authentication cookie, it is a helper cookie for Javascript-based session management. The actual session authentication cookie default name when using Asp.Net Identity is `.AspNetCore.Identity.Application`.

**Comment by MH61Aus** [comment]
I think I'm mostly following here. 

Our current setup is beneath a path, and I'm not sure I'll be able to convince to powers that be to change that. So when you say 

> If IdentityServer answers to different casing of the path, it might issue tokens with different casing to the issuer field which would cause validation problems.

wouldn't the options.LowerCaseIssuerUri handle this? this option defaults to true, so the issuer in my auth token is always lowercase.

So basically, if we are constrained to keeping it setup beneath a path, it seems we just need to tell anyone setting up clients to ensure consistent casing, and the best way to do that would be to always do a to lower.

so for example, in the `AddOpenIdConnect` of each client:
`options.Authority = builder.Configuration["IdentityServerOptions:Authority"]!.ToLowerInvariant();`

And if we want to add a Profile button to any client:

```
var identityManageAccountUrl = identityOptions.Value.Authority!.ToLowerInvariant().AppendPathSegment("/identity/account/manage");

<a class="btn btn-link p-0" href="@identityManageAccountUrl" target="_blank">Profile</a>
```

**Comment by AndersAbel** [comment]
You're right, the `LowerCaseIssuerUri` would remove the risk for issuer mismatch. I overlooked that option when I write the answer.

Yes, you would require all the clients to have the correct casing. As I said above, we do not generally recommend hosting IdentityServer beneath a path but rather on it's own host. One of the things you could do to prevent problems with the cookie path is to create a custom middleware that enforces the right casing on the path. I think that this is one example of where fail-fast is the best option to make sure all clients use the right casing in their redirects.

**Comment by carnahanliam** [comment]
@AndersAbel can you expand on what you mean when you say:

> The security model for cookie isolation works better if IdentityServer is hosted on a separate host.

Is the reason mostly related to the potential issues with cookie path discussed here, or are there other reasons why you would not generally recommend to host the IdentityServer beneath a path?

**Comment by AndersAbel** [comment]
The security model for cookies relies on the host being the security boundary. A cookie with a path will only be *sent* with requests related to that path, but it is still possible for any page/endpoint on the same host to *set* and overwrite cookies on that path. The same is not true for different hosts. Only the specific host can set a cookie that applies to that host.

I would consider path restrictions for cookies to be a performance feature and not primarily a security feature.

**Comment by RolandGuijt** [comment]
Does anybody have anything to add to this issue? If not I'd like to close it.

**Comment by MH61Aus** [comment]
I think we can close this.

in terms of a solution, we've requested a new host so that we can move our Identity server deployment there. In the meantime we're any time we set the authority or generate links to the profile page, we are doing a ToLowerInvariant, and communicating to anyone working on clients that they'll need to follow this convention.

My suggestion is that this is all made more specific in the docs. I know there's always a distinction where we're responsible for knowing how cookies work in .net, but I still don't think I've seen any guidance that states that Duende's recommendation is that the Identity Server is deployed to its own host and not a path.

In the samples, when clients configure the authority, you could potentially add a comment reminding us that it is case sensitive. (Ideally Microsoft's summary tag on the Authority property would warn us that it is case sensitive).

In the AspnetCoreIdentity sample, there's only a secure page writing out the claims. I think one of the huge benefits in using AspNetIdentity is that you get things like the user profile section. I would suggest that in your sample you should show how to link to the profile page from the client, and how to refresh the claims at the client after the profile has been updated - and most importantly reiterate the case sensitivity here, as you'll be presented with a login prompt if you get it wrong.

For us, we acheived this by adding a Account Settings page to the client, which has a link stating that it'll open the profile page in a new tab. The link is built off the authority uri:

```
var identityManageAccountUrl = identityOptions.Value.Authority!.ToLowerInvariant().AppendPathSegment("/identity/account/manage");

<a class="btn btn-link p-0" href="@identityManageAccountUrl" target="_blank">Profile</a>

```

I think seeing something similar to this in the sample would be really helpful.

**Comment by RolandGuijt** [comment]
Thanks for your suggestions!
I will take a look at them and see what we can apply to our docs and samples.
